mlflow/transformers.py

```diff
@@ -1844,9 +1843,90 @@ def _parse_raw_pipeline_input(self, data):
+        can be used for pipeline inference.
+
+        To illustrate the input and outputs of this function, for the following inputs to
+        the pyfunc.predict() call for this pipeline type:
+
+        "text to classify"
+        ["text to classify", "other text to classify"]
+        {"text": "text to classify", "text_pair": "pair text"}
+        [{"text": "text", "text_pair": "pair"}, {"text": "t", "text_pair": "tp" }]
+
+        Pyfunc processing will convert these to the following structures:
+
+        [{0: "text to classify"}]
+        [{0: "text to classify"}, {0: "other text to classify"}]
+        [{"text": "text to classify", "text_pair": "pair text"}]
+        [{"text": "text", "text_pair": "pair"}, {"text": "t", "text_pair": "tp" }]
+
+        The purpose of this function is to convert them into the correct format for input
+        to the pipeline (wrapping as a list has no bearing on the correctness of the
+        inferred classifications):
+
+        ["text to classify"]
+        ["text to classify", "other text to classify"]
+        [{"text": "text to classify", "text_pair": "pair text"}]
+        [{"text": "text", "text_pair": "pair"}, {"text": "t", "text_pair": "tp" }]
+
+        Additionally, for dict input types (the 'text' & 'text_pair' input example), the dict
+        input will be JSON stringified within MLflow model serving. In order to reconvert this
+        structure back into the appropriate type, we use ast.literal_eval() to convert back
+        to a dict. We avoid using JSON.loads() due to pandas DataFrame conversions that invert
+        single and double quotes with escape sequences that are not consistent if the string
+        contains escaped quotes.
+        """
+
+        def _check_keys(payload):
+            """Check if a dictionary contains only allowable keys."""
+            allowable_str_keys = {"text", "text_pair"}
+            if (
+                not all(isinstance(key, int) for key in payload.keys())
+                and set(payload) - allowable_str_keys
+            ):
+                raise MlflowException(
+                    "Text Classification pipelines may only define dictionary inputs with keys "
+                    f"defined as {allowable_str_keys}"
+                )
+
+        if isinstance(data, str):
+            return data
+        elif isinstance(data, dict):
+            _check_keys(data)
+            return data
+        elif isinstance(data, list):
+            if all(isinstance(item, str) for item in data):
+                return data
+            elif all(isinstance(item, dict) for item in data):
+                for payload in data:
+                    _check_keys(payload)
+                if list(data[0].keys())[0] == 0:
+                    data = [list(item.values())[0] for item in data]
+                try:
+                    return [ast.literal_eval(s) for s in data]
```

Comment:
Can we insert a comment on why we're using `literal_eval`?

Reply:
good idea.

Reply:
Can you give me an example of `s`? Curious what it looks like.

Reply:
For the scalar input to MLflow serving of 
```
{
            "inputs": [
                "I think this sushi might have gone off",
                "That gym smells like feet, hot garbage, and sadness",
                "I love that we have a moon",
                "I 'love' debugging subprocesses",
            ]
        }
```
The type casting becomes: 
```
[{0: \'I think this sushi might have gone off\'}, {0: \'That gym smells like feet, hot garbage, and sadness\'}, {0: \'I love that we have a moon\'}, {0: "I \'love\' debugging subprocesses"}]
```

Which with the prior logic of just extracting the values without decoding would create correctness issues with the classification scores (since escaped characters were included in the passed-in string). 

The formatting gets a bit crazy with the dict input though. 

For this input to pyfunc serving:
```
json.dumps({"inputs": [{"text": "a 'b'", "text_pair": "b 'a'"}, {"text": "c 'd'", "text_pair": "d \"c\""}]})
```

The internally parsed structure is:
```
[{0: \'{\\\'text\\\': "a \\\'b\\\'", \\\'text_pair\\\': "b \\\'a\\\'"}\'}, {0: \'{\\\'text\\\': "c \\\'d\\\'", \\\'text_pair\\\': \\\'d "c"\\\'}\'}]
```
Using json.loads() results in a JSONDecodeError due to property names expecting to be enclosed in double quotes rather than single quotes. However, ast.literal_eval is capable of converting the quote structure properly. 

Repro:

```python
json.loads("""[{0: \'{\\\'text\\\': "a \\\'b\\\'", \\\'text_pair\\\': "b \\\'a\\\'"}\'}, {0: \'{\\\'text\\\': "c \\\'d\\\'", \\\'text_pair\\\': \\\'d "c"\\\'}\'}]""")
```
raises

```
---------------------------------------------------------------------------
JSONDecodeError                           Traceback (most recent call last)
/var/folders/cd/n8n0rm2x53l_s0xv_j_xklb00000gp/T/ipykernel_35966/2524061683.py in <cell line: 1>()
----> 1 json.loads("""[{0: \'{\\\'text\\\': "a \\\'b\\\'", \\\'text_pair\\\': "b \\\'a\\\'"}\'}, {0: \'{\\\'text\\\': "c \\\'d\\\'", \\\'text_pair\\\': \\\'d "c"\\\'}\'}]""")

~/miniconda3/envs/mlflow-dev-env/lib/python3.8/json/__init__.py in loads(s, cls, object_hook, parse_float, parse_int, parse_constant, object_pairs_hook, **kw)
    355             parse_int is None and parse_float is None and
    356             parse_constant is None and object_pairs_hook is None and not kw):
--> 357         return _default_decoder.decode(s)
    358     if cls is None:
    359         cls = JSONDecoder

~/miniconda3/envs/mlflow-dev-env/lib/python3.8/json/decoder.py in decode(self, s, _w)
    335 
    336         """
--> 337         obj, end = self.raw_decode(s, idx=_w(s, 0).end())
    338         end = _w(s, end).end()
    339         if end != len(s):

~/miniconda3/envs/mlflow-dev-env/lib/python3.8/json/decoder.py in raw_decode(self, s, idx)
    351         """
    352         try:
--> 353             obj, end = self.scan_once(s, idx)
    354         except StopIteration as err:
    355             raise JSONDecodeError("Expecting value", s, err.value) from None

JSONDecodeError: Expecting property name enclosed in double quotes: line 1 column 3 (char 2)
```

While
```python
ast.literal_eval("""[{0: \'{\\\'text\\\': "a \\\'b\\\'", \\\'text_pair\\\': "b \\\'a\\\'"}\'}, {0: \'{\\\'text\\\': "c \\\'d\\\'", \\\'text_pair\\\': \\\'d "c"\\\'}\'}]""")
```

results in 
```
[{0: '{\'text\': "a \'b\'", \'text_pair\': "b \'a\'"}'},
 {0: '{\'text\': "c \'d\'", \'text_pair\': \'d "c"\'}'}]
```
which, when parsed with the new logic becomes:
```
[{'text': "a 'b'", 'text_pair': "b 'a'"},
 {'text': "c 'd'", 'text_pair': 'd "c"'}]
```

Reply:
@BenWilson2 Where is:

```
json.dumps({"inputs": [{"text": "a 'b'", "text_pair": "b 'a'"}, {"text": "c 'd'", "text_pair": "d \"c\""}]})
```

converted to:

```
[{0: \'{\\\'text\\\': "a \\\'b\\\'", \\\'text_pair\\\': "b \\\'a\\\'"}\'}, {0: \'{\\\'text\\\': "c \\\'d\\\'", \\\'text_pair\\\': \\\'d "c"\\\'}\'}]
```

?

Reply:
in the ML serving code with signature validation logic. To generate these values, I simply inserted a raise at the breakpoints prior to and after the processing logic and embedded the values within the Exception message via interpolation.

Reply:
Thanks for the investigation! I'll dig where the conversion is performed.